### PR TITLE
GA Migrations Java PreCommit PVR Flink Batch and Docker

### PR DESCRIPTION
**Pull request by fernando-wizeline**
As part of the migration of Precommit and Postcommit Jobs from Jenkins to GA in self-hosted runners, this PR contains:

Migrated workflow [job-precommit-java-pvr-flink-batch.yml](https://github.com/fernando-wizeline/beam/blob/gam-precommit-java-pvr-flink-batch-docker/.github/workflows/job-precommit-java-pvr-flink-batch.yml)
Migrated workflow [job-precommit-java-pvr-flink-docker.ymll](https://github.com/fernando-wizeline/beam/blob/gam-precommit-java-pvr-flink-batch-docker/.github/workflows/job-precommit-java-pvr-flink-docker.yml)

The migrated workflow was added to [CI.md](https://github.com/fernando-wizeline/beam/blob/gam-precommit-java-pvr-flink-batch-docker/CI.md)

DO NOT MERGE until the effort to use self-hosted runners is completed https://github.com/apache/beam/pull/22703
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

File: .github/workflows/job-precommit-java-pvr-flink-batch.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+    needs: set-properties
+    name: Run PreCommit Java PVR Flink Batch
+    runs-on: [self-hosted, ubuntu-20.04]
+    strategy:
+      fail-fast: false
+      matrix: #JDK17 execution will fail due to new access policies applied to the core JDK libraries starting at JDK 16.
+        java: ${{fromJson(needs.set-properties.outputs.properties).JavaTestProperties.SUPPORTED_VERSIONS}}
+        flink: ${{fromJson(needs.set-properties.outputs.properties).JavaTestProperties.FLINK_VERSIONS}}
+        exclude: #TODO: Remove exclude and retest once issue https://github.com/apache/beam/issues/23066 gets resolved.
+          - java: "11" #Creation of the DockerEnvironment fails with Java 11.
```
Comment (damccorm) [inline .github/workflows/job-precommit-java-pvr-flink-batch.yml:65]: Is there an issue tracking this?

File: .github/workflows/job-precommit-java-pvr-flink-batch.yml [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  precommit-java-pvr-flink-batch:
+    needs: set-properties
+    name: Run PreCommit Java PVR Flink Batch
+    runs-on: [self-hosted, ubuntu-20.04]
+    strategy:
+      fail-fast: false
+      matrix: #JDK17 execution will fail due to new access policies applied to the core JDK libraries starting at JDK 16.
+        java: ${{fromJson(needs.set-properties.outputs.properties).JavaTestProperties.SUPPORTED_VERSIONS}}
+        flink: ${{fromJson(needs.set-properties.outputs.properties).JavaTestProperties.FLINK_VERSIONS}}
+        exclude: #TODO: Remove exclude and retest once issue https://github.com/apache/beam/issues/23066 gets resolved.
```
Comment (damccorm) [inline .github/workflows/job-precommit-java-pvr-flink-batch.yml:64]: Why does this impact GHA and not Jenkins?

File: .github/workflows/job-precommit-java-pvr-flink-docker.yml [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    needs: set-properties
+    name: Run PreCommit Java PVR Flink Docker
+    runs-on: [self-hosted, ubuntu-20.04]
+    strategy:
+      fail-fast: false
+      matrix: #JDK17 execution will fail due to new access policies applied to the core JDK libraries starting at JDK 16.
+        java: ${{fromJson(needs.set-properties.outputs.properties).JavaTestProperties.SUPPORTED_VERSIONS}}
+        flink: ${{fromJson(needs.set-properties.outputs.properties).JavaTestProperties.FLINK_VERSIONS}}
+        exclude: #TODO: Remove exclude and retest once issue https://github.com/apache/beam/issues/23066 gets resolved.
+          - java: "11" #Creation of the DockerEnvironment fails with Java 11.
```
Comment (damccorm) [inline .github/workflows/job-precommit-java-pvr-flink-docker.yml:67]: Same questions as above

File: .github/workflows/job-precommit-java-pvr-flink-batch.yml (hunk 1 above)
Comment (fernando-wizeline) [reply .github/workflows/job-precommit-java-pvr-flink-batch.yml:65]: No issue for this one yet, I will go ahead and create one.

Comment (fernando-wizeline) [reply .github/workflows/job-precommit-java-pvr-flink-batch.yml:65]: Deleted the previous comment. Just realized the problem with the creation of the DockerEnvironment is due to the ErrorProne library issues when compiled with JDK11. The issue tracking that is this one: https://github.com/apache/beam/issues/23066

File: .github/workflows/job-precommit-java-pvr-flink-batch.yml (hunk 2 above)
Comment (fernando-wizeline) [reply .github/workflows/job-precommit-java-pvr-flink-batch.yml:64]: In jenkins, the jar files are being compiled with JDK8 and executed with JDK11 or 17.
These errors crop up when compiling with JDK11/17.

JDK 9 introduced restrictions on the method visibility for the ByteBuffer class, resulting in NoSuchMethodError exceptions at compile time.
We tried going the same route in GHA, that is, adding compile-time flags in Gradle (--add-exports and --add-opens) to circumvent the ByteBuffer methods visibility restrictions put in place in JDK9, with no success as of yet.

I opened this issue https://github.com/apache/beam/issues/23068 to keep track of this.

JDK16 did something similar but that change affects the ErrorProne library. We also tried applying the same --add-exports and --add-opens flags, but they were not being picked up.

Issue https://github.com/apache/beam/issues/23066 as opened to track that one.

We consulted this with Kerry, and at that moment we decided to keep the associated jenkins jobs and exclude that matrix from the GHA workflows.

**Review by damccorm — APPROVED** [review]

**Pull request closed without merging**